### PR TITLE
Persisting third-party python integrations

### PR DESCRIPTION
**Pull request by jose-manuel-almaza**
### What does this PR do?

Keep non-Core integrations that the user installed for the version they're upgrading. 
The solution must be one that does not imply action by the user (or the minimum possible if no action is impossible). The process of maintaining updated integrations should be transparent to the user.

### Motivation

Updating the Agent wipes all non-Core integrations: Customers continue to express frustration, because each time they upgrade a Datadog Agent, it wipes all non-core integrations, which causes gaps in monitoring, and unnecessary fire drills. This applies to all partner integrations as well as custom checks written by customers, which can be key to a customer’s observability. 

### Additional Notes

RFC: https://docs.google.com/document/d/1b_VAt17c4j6WCyMOfLc_M0qzRFKmyJ1Gq-aHDFP9Dvc/edit?usp=sharing

### Describe how to test/QA your changes

We'll prepare a set of VMs with different OS and PM in order to test this functionality in different scenarios automatically.

### CODE REVIEW

File: omnibus/package-scripts/agent-deb/postinst
Comment (NouemanKHAL) [inline omnibus/package-scripts/agent-deb/postinst:185]: Should we check for the python binary before using it? (same for the python-scripts)

File: omnibus/package-scripts/agent-deb/prerm
Comment (NouemanKHAL) [inline omnibus/package-scripts/agent-deb/prerm:140]: nit: it might be worth to wrap this in a function to keep it readable and self explanatory

File: omnibus/python-scripts/postinst.py
Comment (NouemanKHAL) [inline omnibus/python-scripts/postinst.py:52]: We sort the output to make it easier to generate the diff right?
I think it would be better if we could sort in python as well and not rely on the `sort` utility in case it's not available in the system.

File: omnibus/python-scripts/postinst.py
Comment (NouemanKHAL) [inline omnibus/python-scripts/postinst.py:55]: What happens if `chown` fails? Does it raise an exception?

File: omnibus/package-scripts/agent-deb/postinst
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/postinst:185]: I don't think so. They're shipped along with the agent.

File: omnibus/package-scripts/agent-deb/prerm
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/prerm:140]: I'd prefer keep it simple with just one line. Maybe adding a comment with the purpose?

File: omnibus/python-scripts/postinst.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/postinst.py:52]: Makes sense. I'll rewrite it.

File: omnibus/python-scripts/postinst.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/postinst.py:55]: In this case the python script would fail and the caller (postinst deb script in this case) would get an error

File: omnibus/package-scripts/agent-deb/postinst
Comment (NouemanKHAL) [reply omnibus/package-scripts/agent-deb/postinst:185]: Yes, but we do the same thing for other binaries above in the script:
https://github.com/DataDog/datadog-agent/blob/jose/persisting-python-integrations/omnibus/package-scripts/agent-deb/postinst#L125-L141

Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/postinst:185]: I don't know if they're optional. In any case I can add the check you suggest.

Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/postinst]: Added check

File: releasenotes/notes/fix-persisting-python-integrations-5065a0ded39e76b5.yaml
Comment (drichards-87) [inline releasenotes/notes/fix-persisting-python-integrations-5065a0ded39e76b5.yaml:4]:
```suggestion
    Fix the removal of 'non-core' integrations during Agent upgrades.
```

**Review by drichards-87 — APPROVED** [review]
Left a small suggestion from Docs and approved the PR.

**Comment by jose-manuel-almaza** [comment]
> > We'll prepare a set of VMs with different OS and PM in order to test this functionality in different scenarios automatically.
> 
> When/where is this happening? Could it be included in this PR as an e2e test?

That'd be nice. Where are the current tests for this functionality?

**Comment by jose-manuel-almaza** [comment]
> > We'll prepare a set of VMs with different OS and PM in order to test this functionality in different scenarios automatically.
> 
> When/where is this happening? Could it be included in this PR as an e2e test?

New tests added. Could you pls review them?

File: .gitlab/e2e_install_packages/common.yml
Comment (amenasria) [inline .gitlab/e2e_install_packages/common.yml:60]:
```suggestion
    - DATADOG_AGENT_API_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $INSTALL_SCRIPT_API_KEY ) || exit $?; export DATADOG_AGENT_API_KEY
```

**💬 suggestion:** ‏For now, we need to fetch secret like this to get the proper error to raise in the CI

**Review by amenasria — APPROVED** [review]
Left a comment, otherwise LGTM for DevX Infra as long as the CI passes !

Also please update your branch before merging (it's 20 days i.e ~247 commits behind) ! Thanks !

File: omnibus/package-scripts/agent-deb/prerm
Comment (chouquette) [inline omnibus/package-scripts/agent-deb/prerm:162]: Do we need to run this when we're uninstalling? It seems only relevant when upgrading no?

File: .gitlab/e2e_install_packages/common.yml
Comment (NouemanKHAL) [reply .gitlab/e2e_install_packages/common.yml]: Nice catch @amenasria ! I remember seeing a PR about that, I guess I copied the old changes because the branch is behind.
However, we don't see the new-e2e test on the CI? When are the upgrade tests executed?

File: omnibus/python-scripts/packages.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/packages.py:50]: I don't see it as a big improvement, I think the current code is much cleaner (by looking at the calls from the client code we can easily tell what's going on there). Also if in the future we have to change the name of a file(s), we would only have to change it in one place.
In this case I think less code would not mean cleaner code.

File: omnibus/python-scripts/packages.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/packages.py:161]: The option you mention is also good but I don't see how it improves the current code.

File: omnibus/package-scripts/agent-rpm/prerm
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-rpm/prerm]: Good catch! Fixed!

File: omnibus/package-scripts/agent-deb/prerm
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/prerm]: Good catch! Fixed!

File: omnibus/python-scripts/packages.py
Comment (chouetz) [reply omnibus/python-scripts/packages.py:97]: You have 2 methods calling `print` + `run_command`. So I suggest the `print` to be inside the `run_command` method: as a consequence you don't need the 2 intermediate methods you created. 
Like above this is not a major improvement, it's just me not being a big fan of duplicated code. You can keep it like that.

File: omnibus/python-scripts/packages.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/packages.py]: Good catch! Fixed!

File: .gitlab/e2e_install_packages/common.yml
Comment (NouemanKHAL) [reply .gitlab/e2e_install_packages/common.yml:54]: Addressed in #29760

Comment (NouemanKHAL) [reply .gitlab/e2e_install_packages/common.yml:54]: Addressed in #29760

Comment (NouemanKHAL) [reply .gitlab/e2e_install_packages/common.yml:54]: Addressed in #29760

File: .gitlab/e2e_install_packages/common.yml
Comment (NouemanKHAL) [reply .gitlab/e2e_install_packages/common.yml]: Addressed in #29760

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (NouemanKHAL) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:121]: We can delete it, no big deal, I just left it there to discuss it.

**Comment by NouemanKHAL** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by NouemanKHAL** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by NouemanKHAL** [comment]
Note: I accidentally clicked twice on the `Run all builds and tests` button, so I canceled the 2nd pipeline 45728943.

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [inline test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:147]: **❓ question**
Have you considered using `host.WriteFile` instead? 

https://github.com/DataDog/datadog-agent/blob/main/test/new-e2e/pkg/utils/e2e/client/host.go#L250-L264

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:88]: If we mark all sub-tests, I'd rather mark it at parent test context, at the beginning of `TestPersistingIntegrations`

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (NouemanKHAL) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:147]: Yes, I got a `Permission Denied` error, I had the same error with `host.CopyFile`

**Comment by pducolin** [comment]
Note: as the test is marked as flake, it will not run on the dev pipelines

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [inline test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:5]: **❓ question**
Why is this inside `agent-platform`? Would have assumed it would be in `test/new-e2e/tests/persisting-integrations`

Comment (NouemanKHAL) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:5]: We did that because in the beginning we wanted to add a test to the existing `test/new-e2e/tests/agent-platform/upgrade` test. But we couldn't change it, so we shared the codebase and added it next to it.

**Comment by NouemanKHAL** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by NouemanKHAL** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:5]: Ok, possible follow up: migrating common code to `pkg` and move up these tests - not urgent

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:172]: It's probably executed without sudo, thank you for the context!

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [inline test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:92]: **💬 suggestion**
I would move the skip up here:

```suggestion
	flake.Mark(t)
	platformJSON := map[string]map[string]map[string]string{}
```

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (pducolin) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:113]: Would remove the marking here
```suggestion
```

**Review by pducolin — APPROVED** [review]

File: test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go
Comment (NouemanKHAL) [reply test/new-e2e/tests/agent-platform/persisting-integrations/persisting_integrations_test.go:94]: Done!

**Comment by pducolin** [comment]
Thank you for addressing the comments!

**Review by NouemanKHAL — APPROVED** [review]

File: omnibus/python-scripts/postinst.py
Comment (chouquette) [inline omnibus/python-scripts/postinst.py:21]: When/why is this file supposed to be created? It doesn't seem to be written to nor read from in your PR

File: omnibus/python-scripts/packages.py
Comment (chouquette) [reply omnibus/python-scripts/packages.py:163]: Given that `cleanup_files` only remove 2 files at most (in the current setup at least) I'd also find it easier to reason about the code if it calls into a standard function.
Seeing a call to `packages.cleanup_files` I tend to assume that the function will do something more specific than just removing a file

File: omnibus/python-scripts/packages.py
Comment (chouquette) [inline omnibus/python-scripts/packages.py:93]: I suppose this could also be `chown`ed to `dd-agent` for consistency with files created by `create_python_installed_packages_file`

File: omnibus/python-scripts/postinst.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/postinst.py:21]: This file is our feature flag.
Using a file as a flag offers greater persistence and flexibility compared to environment variables (the first option we thought), as it remains accessible across sessions and subprocesses, including during apt-get installs. Files are also easier to manage, allowing manual or programmatic creation and removal, and are more traceable for debugging. This approach avoids the risk of environment variables not being propagated correctly, making it more reliable in isolated environments.
You can see we're creating this file in our tests.

Comment (NouemanKHAL) [reply omnibus/python-scripts/postinst.py:21]: @chouquette This file is acting like a feature flag, it's supposed to be created by users willing to use try this upgrade feature.
We couldn't use an environment variable for various reasons...

Comment (chouquette) [reply omnibus/python-scripts/postinst.py:21]: Gotcha, thanks for the explanation!

File: omnibus/python-scripts/packages.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/packages.py:94]: Fixed! Good catch!

File: omnibus/package-scripts/agent-rpm/prerm
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-rpm/prerm]: Good catch! Copy&Paste effect I guess
Fixed!

File: omnibus/package-scripts/agent-deb/prerm
Comment (jose-manuel-almaza) [reply omnibus/package-scripts/agent-deb/prerm:161]: Exactly. That's the reason. I'll add a comment as you suggest

File: omnibus/python-scripts/packages.py
Comment (jose-manuel-almaza) [reply omnibus/python-scripts/packages.py:164]: In my opinion, implementing file deletion in this function simplifies the caller's code and equalizes levels of abstraction. It also makes it easier to add more specific tasks in the future, as you mentioned.

File: releasenotes/notes/fix-persisting-python-integrations-5065a0ded39e76b5.yaml
Comment (chouquette) [inline releasenotes/notes/fix-persisting-python-integrations-5065a0ded39e76b5.yaml:4]: Given this is behind a flag, should it be advertised in release notes now?

**Review by chouquette — APPROVED** [review]

File: omnibus/python-scripts/packages.py
Comment (chouetz) [reply omnibus/python-scripts/packages.py:164]: My point was to save the `if file exists` using a dedicated method. As other similar comments I did, that's just sharing opinion about code syntax. Your current way is fine

**Review by chouetz — APPROVED** [review]

File: test/new-e2e/tests/agent-platform/persisting-integrations/fixtures/diff.txt
Comment (NouemanKHAL) [reply test/new-e2e/tests/agent-platform/persisting-integrations/fixtures/diff.txt]: I don't see where this message is displayed but I just added a newline, I guess that should fix it.

**Pull request merged**
